state/store.go

```diff
@@ -366,6 +389,51 @@ func (store dbStore) PruneStates(from int64, to int64, evidenceThresholdHeight i
 	return nil
 }
 
+// PruneABCIResponses attempts to prune all ABCI responses up to, but not
+// including, the given height. On success, returns the height to which ABCI
+// responses were pruned.
+func (store dbStore) PruneABCIResponses(height int64) (uint64, error) {
```

Comment:
@glnro and I were talking about this method today and realized we should refactor it a bit. There are two main changes I've implemented here that I'd appreciate your perspective on @jmalicevic:

1. The pruned height counting mechanism's been refactored.
2. We're now caching the last ABCI responses pruned height. I imagine that in chains at large heights (e.g. over 1 million), this would be substantially sub-optimal to run through all possible heights for every pruning operation. @glnro recommended supplying a `fromHeight` and `toHeight` parameter to this method instead of just `height`, but thinking about it now, under what circumstances would we want to be able to offer pruning of selective chunks of ABCI responses? Seems to me as though, if we want to prune up to a particular height, all of the relevant mechanisms for that pruning should be self-contained.

Reply:
That being said there could be a use case for selectively pruning a range but I imagine this would be for extenuating circumstances (debugging). Conceivably we could pass a `fromHeight` and always call it with the latest pruned height.

Reply:
Hey so here is my take:
1. I think it's a good way to improve the count and we should improve the same for counting the blocks pruned. 
2. We were always caching it but not in a persistent way. I just saw that the refactored code was not actually updating the `lastABCIResPrunedHeight`, but with that fixed, the pruner always remembers the height up to which it pruned the ABCI results and prunes again, only if the retain height is different from the last height up to which we pruned. And we always prune from height 1 to `retainHeight`. Which means that `lastABCIResPrunedHeight < retainHeight` for all correct cases. 
That being said, if pruning works correctly, line 414 will error once we hit a height (which is height 1 right now) that does not exist and we should indeed always be pruning backwards from `retainHeight` until either: 
 a) We hit an error , in which we should flush the batch up until that point (which we are not doing now)
 b) We pass `lastABCIResPrunedHeight` as a `fromHeight` and prune between that and the `retainHeight`. If this is not persistent though it is not the best way. 

I do not see a use case where we should partially delete the responses (between two heights). It should always be from 1 to `retainHeight` and then from there upwards. Otherwise we open ourselves for a bunch of corner cases. That is why I would not pass a second parameter to the function, but rather go with option `a)` and check for a `ErrKeyNotFound` on delete, and return. Or we persist the parameter and check for it internally, without passing it as a function parameter.

Reply:
To sum up, the idea was always to cache this but the old way of caching would actually not work because the `Delete` statement would always error after the initial pruning is done. So we either keep that way but prune backwards until we error or do it the way you are doing it now with persisting the info but not passing it as a `fromHeight` parameter ,rather reading it within the function. 
For block pruning, there is no reason to persist this because we implicitly do this by updating `store.Base()` on pruning which is always checked so we have this info.

Reply:
The backwards pruning seems more optimal to me. If we go that route, and we get an error, is it a specific kind of error that we can detect like `ErrKeyNotFound`? That would then do away with the need for tracking the last retain height in the DB and we could revert those changes of mine.

Reply:
I believe yes but I can do a quick check. If that is the case, I'll push a version without persisting the last height in the DB.